### PR TITLE
Add code mirror editors to the dev-tools data fields

### PR DESCRIPTION
**Pull request by bramkragten**
image

image

### CODE REVIEW

File: src/components/ha-code-editor.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  set rtl(rtl: boolean) {
+    this._rtl = rtl;
+    this.setScrollBarDirection();
+  }
+
+  set autofocus(autofocus: boolean) {
+    this._autofocus = autofocus;
+    if (this.codemirror) {
+      this.codemirror.focus();
```
Comment (balloob) [inline src/components/ha-code-editor.ts]: This is not what I expect `autofocus` to do. Docs: https://developer.mozilla.org/en-US/docs/Web/API/HTMLSelectElement/autofocus

If we want to stay close to how HTML works, let's check for `autofocus` inside `connectedCallback` and focus if it's true.

File: src/components/ha-code-editor.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+export class HaCodeEditor extends HTMLElement {
+  public codemirror?: any;
+  private mode: string;
+  private _autofocus = false;
+  private _rtl = false;
+  private _value: string;
+
+  public constructor(mode: string) {
```
Comment (balloob) [inline src/components/ha-code-editor.ts]: Why would we make this a constructor and not just another property?

File: src/components/ha-code-editor.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
+// @ts-ignore
+import CodeMirror from "codemirror";
```
Comment (balloob) [inline src/components/ha-code-editor.ts]: I think that we should async load both this and the CSS (as 1 module with 1 `import()`).

File: src/components/ha-code-editor.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+// @ts-ignore
+import codeMirrorCSS from "codemirror/lib/codemirror.css";
+import { fireEvent } from "../common/dom/fire_event";
+
+declare global {
+  interface HASSDomEvents {
+    "editor-save": undefined;
```
Comment (balloob) [inline src/components/ha-code-editor.ts]: Where is this fired ?

File: src/components/ha-code-editor.ts (hunk 2 above)
Comment (balloob) [reply src/components/ha-code-editor.ts]: And let's just not deal with it being updated

File: src/panels/developer-tools/template/developer-tools-template.js [hunk 5]
```diff
@@ -85,11 +85,13 @@ class HaPanelDevTemplate extends PolymerElement {
...
-          <paper-textarea
-            label="Template editor"
-            value="{{template}}"
-            autofocus
-          ></paper-textarea>
+          <p>Template editor</p>
+          <ha-jinja-editor
+            value="[[template]]"
+            error="[[error]]"
+            autofocus="[[true]]"
```
Comment (balloob) [inline src/panels/developer-tools/template/developer-tools-template.js]: does this work ?

File: src/resources/codemirror.ondemand.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+interface LoadedCodeMirror {
+  codeMirror: any;
+  codeMirrorCss: any;
+}
+
+let loaded: Promise<LoadedCodeMirror>;
+
+export const loadCodeMirror = async (): Promise<LoadedCodeMirror> => {
+  if (!loaded) {
+    loaded = import(/* webpackChunkName: "codemirror" */ "./codemirror");
```
Comment (balloob) [inline src/resources/codemirror.ondemand.ts]: I know we've done it like this before, but thinking now if WebPack caches this import call ?

File: src/resources/codemirror.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+// @ts-ignore
+import _CodeMirror from "codemirror";
+// @ts-ignore
+import _codeMirrorCss from "codemirror/lib/codemirror.css";
+import "codemirror/mode/yaml/yaml";
```
Comment (balloob) [inline src/resources/codemirror.ts]: how big are these files ? Is it ok to bundle all synax up in 1 file ?

**Review by balloob — APPROVED** [review]

File: src/resources/codemirror.ts (hunk 7 above)
Comment (bramkragten) [reply src/resources/codemirror.ts]: Yeah, they are small

**Pull request merged**